src/Functions/FunctionsJSON.h

```diff
@@ -276,37 +278,139 @@ class FunctionJSONHelpers
 
 
 template <typename Name, template<typename> typename Impl>
-class FunctionJSON : public IFunction, WithContext
+class ExecutableFunctionJSON : public IExecutableFunction, WithContext
 {
+
 public:
-    static FunctionPtr create(ContextPtr context_) { return std::make_shared<FunctionJSON>(context_); }
-    FunctionJSON(ContextPtr context_) : WithContext(context_) {}
+    explicit ExecutableFunctionJSON(const NullPresence & null_presence_, bool allow_simdjson_)
+        : null_presence(null_presence_), allow_simdjson(allow_simdjson_)
+    {
+    }
 
-    static constexpr auto name = Name::name;
     String getName() const override { return Name::name; }
-    bool isVariadic() const override { return true; }
-    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForNulls() const override { return false; }
     bool useDefaultImplementationForConstants() const override { return true; }
-    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
-
-    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
-    {
-        return Impl<DummyJSONParser>::getReturnType(Name::name, arguments);
-    }
 
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
     {
+        if (null_presence.has_null_constant)
+            return result_type->createColumnConstWithDefaultValue(input_rows_count);
+
+        ColumnsWithTypeAndName temporary_columns;
+        if (null_presence.has_nullable)
+            temporary_columns = createBlockWithNestedColumns(arguments);
+        else
+            temporary_columns = arguments;
+
+        ColumnPtr temporary_result;
+
         /// Choose JSONParser.
 #if USE_SIMDJSON
-        if (getContext()->getSettingsRef().allow_simdjson)
-            return FunctionJSONHelpers::Executor<Name, Impl, SimdJSONParser>::run(arguments, result_type, input_rows_count);
+        if (allow_simdjson)
+            temporary_result = FunctionJSONHelpers::Executor<Name, Impl, SimdJSONParser>::run(
+                temporary_columns, result_type, input_rows_count);
+        else
 #endif
-
+        {
 #if USE_RAPIDJSON
-        return FunctionJSONHelpers::Executor<Name, Impl, RapidJSONParser>::run(arguments, result_type, input_rows_count);
+            temporary_result = FunctionJSONHelpers::Executor<Name, Impl, RapidJSONParser>::run(
+                temporary_columns, result_type, input_rows_count);
 #else
-        return FunctionJSONHelpers::Executor<Name, Impl, DummyJSONParser>::run(arguments, result_type, input_rows_count);
+            temporary_result
+                = FunctionJSONHelpers::Executor<Name, Impl, DummyJSONParser>::run(temporary_columns, result_type, input_rows_count);
 #endif
+        }
```

Comment:
Let's move this code into a separate function `chooseJSONParserAndExecute()`?

Reply:
Sure.